### PR TITLE
Add right panel chat timeline

### PR DESCRIPTION
**Pull request by toger5**
Fixes: https://github.com/vector-im/element-web/issues/19633
This adds a new possible card for the right panel: `TimelineCard`
 - This card shows the timeline and allows to send messages
 - When pressing on a thread `EventTile`, it shows the thread. The back arrow brings you back to the timeline.
     - This is executed by providing a new state parameter `previousPhase` for the right panel that is updated using the `AfterRightPanelPhaseChangePayload` dispatch event payload.

This does not include:
 - only showing the rightPanelTimeline in rooms that have a maximised widget: https://github.com/vector-im/element-web/issues/19634
 - showing the correct icons (only core navigation) if a room has a maximised widget:https://github.com/vector-im/element-web/issues/19632
This should be merged after: https://github.com/matrix-org/matrix-react-sdk/pull/7100
Screenshot from 2021-11-10 19-43-49

---
Here's what your changelog entry will look like:

## ✨ Features
 * Add right panel chat timeline ([\#7112](https://github.com/matrix-org/matrix-react-sdk/pull/7112)). Fixes vector-im/element-web#19633 and vector-im/element-web#19633. Contributed by @toger5.

Preview: https://61a4d397366dc64538913ff2--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.

### CODE REVIEW

File: src/stores/widgets/WidgetLayoutStore.ts [hunk 1]
```diff
@@ -452,6 +456,11 @@ export class WidgetLayoutStore extends ReadyWatchingStore {
                 for (const w of this.getContainerWidgets(room, Container.Center)) {
                     this.moveToContainer(room, w, Container.Right);
                 }
+                // Additionally change the right panel phase to chat:
+                dis.dispatch<SetRightPanelPhasePayload>({
+                    action: Action.SetRightPanelPhase,
+                    phase: RightPanelPhases.TimelineCard,
+                });
```
Comment (toger5) [inline src/stores/widgets/WidgetLayoutStore.ts:463]: Should we put this inside one of the `onWidgetLayoutChange` callbacks to keep ui logic out of the store? Or is this fine, since its just a dispatch? (cannot fail in other environments like the WidgetLayoutStore-test)

File: src/components/views/right_panel/TimelineCard.tsx [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+    public render(): JSX.Element {
+        return (
+            <BaseCard
+                className="mx_ThreadPanel mx_TimelineCard"
+                onClose={this.props.onClose}
+                withoutScrollContainer={true}
+                header={this.renderTimelineCardHeader()}>
+                <TimelinePanel
+                    showReadReceipts={true}
```
Comment (jryans) [inline src/components/views/right_panel/TimelineCard.tsx:74]: Hmm, I am puzzled by this one being `true`... Weren't we just discussing it's unclear how to display them, so we would leave it off for now?

Comment (toger5) [reply src/components/views/right_panel/TimelineCard.tsx:74]: The main concern was that I though it was not easily possible to add this. After the discussion and fixing the error with the optional callback I tried it out and it seemed to behave really well. Thats why I decided to keep it for review to see what you and Gaelle thinks. 
Other than the main timeline the read receipts don't get a dedicated space but just overlap with the message so it does not impact the width of the text and looks really usable:
main timline                                         |                                     right card timeline
image

File: src/components/views/rooms/EventTile.tsx [hunk 3]
```diff
@@ -619,11 +621,9 @@ export default class EventTile extends React.Component<IProps, IState> {
...
                 className="mx_ThreadInfo"
                 onClick={() => {
-                    dispatchShowThreadEvent(
-                        this.props.mxEvent,
-                    );
-                }}
-            >
+                    dispatchShowThreadEvent(this.props.mxEvent);
+                }}>
+                <span className="mx_ThreadInfo_thread-icon" />
```
Comment (jryans) [inline src/components/views/rooms/EventTile.tsx:626]: What's happening here...? Perhaps a confused merge from upstream thread work?

File: src/components/views/right_panel/TimelineCard.tsx [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        </div>;
+    };
+
+    public render(): JSX.Element {
+        return (
+            <BaseCard
+                className="mx_ThreadPanel mx_TimelineCard"
+                onClose={this.props.onClose}
+                withoutScrollContainer={true}
+                header={this.renderTimelineCardHeader()}>
```
Comment (jryans) [inline src/components/views/right_panel/TimelineCard.tsx:72]:
```suggestion
                header={this.renderTimelineCardHeader()}
            >
```

File: src/components/structures/RightPanel.tsx [hunk 5]
```diff
@@ -348,7 +355,8 @@ export default class RightPanel extends React.Component<IProps, IState> {
                     initialEvent={this.state.initialEvent}
                     initialEventHighlighted={this.state.initialEventHighlighted}
                     permalinkCreator={this.props.permalinkCreator}
-                    e2eStatus={this.props.e2eStatus} />;
+                    e2eStatus={this.props.e2eStatus}
+                    previousPhase={RightPanelStore.getSharedInstance().previousPhase} />;
```
Comment (jryans) [inline src/components/structures/RightPanel.tsx:359]: Hmm, it feels a bit odd to pass the store's value as a prop like this... Perhaps `ThreadView` can ask the store for this value internally?

File: src/components/views/elements/AppTile.tsx [hunk 6]
```diff
@@ -406,6 +409,11 @@ export default class AppTile extends React.Component<IProps, IState> {
             ? Container.Right
             : Container.Center;
         WidgetLayoutStore.instance.moveToContainer(this.props.room, this.props.app, targetContainer);
+        if (targetContainer === Container.Right
+            && RightPanelStore.getSharedInstance().visibleRoomPanelPhase === RightPanelPhases.TimelineCard) {
+            // If the widget gets closed also close the RightPanel chat.
+            dis.dispatch({ action: Action.SetRightPanelPhase, phase: RightPanelPhases.RoomSummary });
```
Comment (jryans) [inline src/components/views/elements/AppTile.tsx:415]: Hmm, it might be more natural to close the right panel entirely, but I'll have to think about it more.

File: src/components/views/right_panel/RoomHeaderButtons.tsx [hunk 7]
```diff
@@ -141,6 +163,12 @@ export default class RoomHeaderButtons extends HeaderButtons<IProps> {
                 isHighlighted={this.isPhase(RightPanelPhases.PinnedMessages)}
                 onClick={this.onPinnedMessagesClicked}
             />
+            <TimelineCardHeaderButton
```
Comment (jryans) [inline src/components/views/right_panel/RoomHeaderButtons.tsx:166]: At the moment, this seems to show the header button in every room. It should only appear in rooms with a maximised widget.

**Review by jryans — CHANGES_REQUESTED** [review]
Thanks, seems like a good start overall! I'll take another look tomorrow, but wanted to send at least some initial feedback.

File: src/components/structures/RightPanel.tsx (hunk 5 above)
Comment (toger5) [reply src/components/structures/RightPanel.tsx:359]: This is a good idea. In the beginning I wanted to keep the current bahaviour of the back button of the TreadView always going back to the threadPanel except if the thread view was opened from the timlelineCard. This is why I also had the previousPhase property in the EventTile. I would have needed to know where the "thread open request" originated from, to decided what the back button does. But since I new went with this simpler approach it is possible to just ask for the previousPhase in the threads view.
This however does change the behaviour!
RoomSummary is open -> click on a thread in the timeline -> click the back button in the ThreadView -> the RoomSummary will open (I think) before this opend the threadPanel

File: src/components/views/elements/AppTile.tsx (hunk 6 above)
Comment (toger5) [reply src/components/views/elements/AppTile.tsx:415]: When testing I also found myself expecting this. It is nice though to at least see the widgets list after pressing minise on the app Tile. Otherwise there is nothing related to widgets on the screen which I would not expect when pressing a minimise button. Than a close button would be more reasonable?

File: src/components/views/right_panel/RoomHeaderButtons.tsx (hunk 7 above)
Comment (toger5) [reply src/components/views/right_panel/RoomHeaderButtons.tsx:166]: This is handled in: https://github.com/matrix-org/matrix-react-sdk/pull/7114
I thought it makes sense to bundle that logic in the context of the other PR

File: src/components/views/elements/AppTile.tsx [hunk 8]
```diff
@@ -406,6 +409,11 @@ export default class AppTile extends React.Component<IProps, IState> {
             ? Container.Right
             : Container.Center;
         WidgetLayoutStore.instance.moveToContainer(this.props.room, this.props.app, targetContainer);
+        if (targetContainer === Container.Right
+            && RightPanelStore.getSharedInstance().visibleRoomPanelPhase === RightPanelPhases.TimelineCard) {
```
[the remark this answers is not in the document]
Comment (toger5) [reply src/components/views/elements/AppTile.tsx:413]: I decided to not include that in this PR. The things not included are listed in the PR description.
> This does not include:
>  - only showing the rightPanelTimeline in rooms that have a maximised widget

We can however just hide them for now. But this will be overhauled when working on https://github.com/vector-im/element-web/issues/19632 anyways.

File: src/components/views/right_panel/TimelineCard.tsx (hunk 2 above)
Comment (toger5) [reply src/components/views/right_panel/TimelineCard.tsx:75]: Should I disable them?
Or can we try them since they of course have their purpose and are not that broken even with smaller timeline widths.
Maybe it is even a good reference for the design team to see what can be improoved?
 </img>
^ I finally manged to load enough read receipts in my test account to have a better example.

Comment (toger5) [reply src/components/views/right_panel/TimelineCard.tsx:75]: There are cases where it can overlap with the text though. Which maybe is a good reason to disable it for now

File: src/components/views/rooms/EventTile.tsx [hunk 9]
```diff
@@ -322,6 +323,7 @@ interface IProps {
     showThreadInfo?: boolean;
 
     timelineRenderingType?: TimelineRenderingType;
+    threadViewPreviousCard?: RightPanelPhases;
```
[the remark this answers is not in the document]
Comment (toger5) [reply src/components/views/rooms/EventTile.tsx]: It could just be removed!

File: src/components/structures/RightPanel.tsx (hunk 5 above)
Comment (toger5) [reply src/components/structures/RightPanel.tsx]: I now fixed it with some additional logic to only allow the back button to bring you to the threads panel or the timeline.
See: [`d6ad1ae` (#7112)](https://github.com/matrix-org/matrix-react-sdk/pull/7112/commits/d6ad1aeb3439e293684cd863cbff7b9dfb0bb96f)

File: src/components/structures/ThreadView.tsx [hunk 10]
```diff
@@ -203,6 +204,15 @@ export default class ThreadView extends React.Component<IProps, IState> {
             event_id: this.state.thread?.id,
         };
 
+        let previousPhase = RightPanelStore.getSharedInstance().previousPhase;
+        // Make sure the previous Phase is always one of the two: Timeline or ThreadPanel
+        if (![RightPanelPhases.ThreadPanel, RightPanelPhases.Timeline].includes(previousPhase)) {
```
Comment (jryans) [inline src/components/structures/ThreadView.tsx:209]: As discussed, we should probably guard this with the widgets feature flag, so there's no functional change for threads work by itself.

File: src/components/views/elements/AppTile.tsx [hunk 11]
```diff
@@ -406,6 +409,11 @@ export default class AppTile extends React.Component<IProps, IState> {
             ? Container.Right
             : Container.Center;
```
Comment (jryans) [inline src/components/views/elements/AppTile.tsx:410]: Please indent the 3 lines below `targetContainer =` by one step. A bit surprised linting did not yell about this...

File: src/components/views/elements/AppTile.tsx [hunk 12]
```diff
@@ -406,6 +409,11 @@ export default class AppTile extends React.Component<IProps, IState> {
...
         WidgetLayoutStore.instance.moveToContainer(this.props.room, this.props.app, targetContainer);
+        if (targetContainer === Container.Right
+            && RightPanelStore.getSharedInstance().visibleRoomPanelPhase === RightPanelPhases.Timeline) {
```
Comment (jryans) [inline src/components/views/elements/AppTile.tsx:413]:
```suggestion
        if (
            targetContainer === Container.Right &&
            RightPanelStore.getSharedInstance().visibleRoomPanelPhase === RightPanelPhases.Timeline
        ) {
```

File: src/components/views/elements/AppTile.tsx (hunk 6 above)
Comment (jryans) [reply src/components/views/elements/AppTile.tsx:415]: Hmm hmm, still not quite sure here. Maybe we leave it like this for now, and then adjust as we test it out more.

File: src/components/views/right_panel/RoomHeaderButtons.tsx (hunk 7 above)
Comment (jryans) [reply src/components/views/right_panel/RoomHeaderButtons.tsx:166]: Ah okay, seems okay to bundle it that way if you want to.

File: src/components/views/right_panel/TimelineCard.tsx (hunk 2 above)
Comment (jryans) [reply src/components/views/right_panel/TimelineCard.tsx:74]: Yes, I think that's enough evidence that design tweaks are needed for them to work well in a small space, and thus we should disable them for now.

File: src/components/views/rooms/EventTile.tsx [hunk 13]
```diff
@@ -619,11 +619,8 @@ export default class EventTile extends React.Component<IProps, IState> {
             <div
                 className="mx_ThreadInfo"
                 onClick={() => {
-                    dispatchShowThreadEvent(
-                        this.props.mxEvent,
-                    );
-                }}
-            >
+                    dispatchShowThreadEvent(this.props.mxEvent);
+                }}>
```
Comment (jryans) [inline src/components/views/rooms/EventTile.tsx:623]: This seems unrelated, should it be reverted...?

File: src/components/views/elements/AppTile.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (jryans) [reply src/components/views/elements/AppTile.tsx]: Hmm, after testing it again today, I think we should at least change the right panel phase when entering a room without maximised widgets as an initial approach. It's quite confusing / disturbing to see two timelines on screen together, so let's ensure that is not possible.

**Review by jryans — CHANGES_REQUESTED** [review]
Thanks, getting closer! 😄 

I expect there may be a few more tweaks, especially as we learn more about right panel navigation.

File: src/components/views/elements/AppTile.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (toger5) [reply src/components/views/elements/AppTile.tsx]: I came to the conclusion that as well. Sadly took me a while today (I wanted it to be a little change since it needs to be changed again when we overhaul the right-panel navigation)... but now its here: [`f8ae27e` (#7112)](https://github.com/matrix-org/matrix-react-sdk/pull/7112/commits/f8ae27ee1857ec5bfddf3dd5812f72b9a8391178) now.

File: src/components/views/rooms/EventTile.tsx (hunk 13 above)
Comment (toger5) [reply src/components/views/rooms/EventTile.tsx:623]: I prefer how it looks with those changes. But you are right that it does not fit into this pr. I guess I should revert it.

File: src/components/views/elements/AppTile.tsx (hunk 6 above)
Comment (toger5) [reply src/components/views/elements/AppTile.tsx:415]: This solved itself. the way I close the timeline for now (before we have per room right panel state) interferes with this. So I remove it for now and the standard behavior now is: maximised widget + right panel chat -(on widget close)-> nothing open.

File: src/components/structures/RoomView.tsx [hunk 14]
```diff
@@ -354,6 +355,20 @@ export class RoomView extends React.Component<IRoomProps, IRoomState> {
             : MainSplitContentType.Timeline;
     };
 
+    private checkRightPanel = (room) => {
+        // This is a hack to hide the chat. This should not be necassary once the right panel
+        // phase is stored per room. (need to be done after check widget so that mainSplitContentType is updated)
+        if (RightPanelStore.getSharedInstance().roomPanelPhase === RightPanelPhases.Timeline
+            && this.state.showRightPanel
+            && !WidgetLayoutStore.instance.hasMaximisedWidget(this.state.room)) {
```
Comment (jryans) [inline src/components/structures/RoomView.tsx:363]:
```suggestion
        if (
            RightPanelStore.getSharedInstance().roomPanelPhase === RightPanelPhases.Timeline &&
            this.state.showRightPanel &&
            !WidgetLayoutStore.instance.hasMaximisedWidget(this.state.room)
        ) {
```

File: src/components/structures/RoomView.tsx [hunk 15]
```diff
@@ -354,6 +355,20 @@ export class RoomView extends React.Component<IRoomProps, IRoomState> {
             : MainSplitContentType.Timeline;
     };
 
+    private checkRightPanel = (room) => {
+        // This is a hack to hide the chat. This should not be necassary once the right panel
```
Comment (jryans) [inline src/components/structures/RoomView.tsx:359]:
```suggestion
        // This is a hack to hide the chat. This should not be necessary once the right panel
```

**Review by jryans — CHANGES_REQUESTED** [review]
Just a few small things, almost ready to merge! 😄

File: src/components/structures/RoomView.tsx (hunk 14 above)
Comment (toger5) [reply src/components/structures/RoomView.tsx]: Oh I forgot this one... you commented that earlier already.

**Review by jryans — APPROVED** [review]
Looks great, thanks for working on this! 😄

**Review by gaelledel — CHANGES_REQUESTED** [review]
It seems the behaviour for when Threads is enabled hasn't been implemented.
See the mockups here https://www.figma.com/file/k2bzeSQmUipBd5vGsTzMUl/Maximised-widgets?node-id=130%3A20948
The Threads button should disappear when entering the maximised widget layout
It should only be possible to view a specific thread within the chat panel
When users click on the thread, it leads to that specific thread only
Users can go back, to see the core chat again

**Comment by gaelledel** [comment]
Screenshot 2021-11-24 at 15 25 06
I wonder if it'd be possible to tailor the chat layout a little bit more? There is a lot of empty space (see marks on image).

**Comment by toger5** [comment]
The gap between the two widgets need to be revisited because otherwise we run into issues with the scroll bar. This might be a little bit more effort because the widget container should also have a gap to the right  window when the right panel is closed. Is gap between the widget container and the right panel a blocker? Or just a nice to have?
I can try to get the area for the profile pictures smaller however.

**Comment by toger5** [comment]
The profile picture spacing is very entangled with how the threads view (that also should be the case, so that the right panel timeline is similar to the thread panel) I think this change should also be done in a separate PR and with input from the threads team.

**Comment by toger5** [comment]
I tracked the mentioned issues: https://github.com/vector-im/element-web/issues/19905 and https://github.com/vector-im/element-web/issues/19904

**Comment by gaelledel** [comment]
> @gaelledel I tracked the mentioned issues: [vector-im/element-web#19905](https://github.com/vector-im/element-web/issues/19905) and [vector-im/element-web#19904](https://github.com/vector-im/element-web/issues/19904)

Sorry. I am unable to see a preview of the work

**Comment by toger5** [comment]
What preview do you mean?
This pr is previewed here: 619e5fc067f49471c2699ff2--matrix-react-sdk.netlify.app
The links i sended were issues to track the stuff that is not part of this pr. (but I havent implemented that yet)

**Review by gaelledel — CHANGES_REQUESTED** [review]
- [ ] I will synch with @janogarcia (designer on threads project) regarding the spacing tweaks for the panel

- [ ] Could you remove the overflow menu on the panel nav?
Screenshot 2021-11-29 at 11 29 32
- [ ] Instead please add an overflow menu and only The **"copy link to thread option"** on hover on the thread message. Like below
Screenshot 2021-11-29 at 11 32 05

- [ ] Can anything be done regarding the issue below? The timestamp overlaps the chevron. Perhaps this is for @janogarcia to discuss with the threads team.
Screenshot 2021-11-29 at 11 35 18

**Comment by toger5** [comment]
the issues @gaelledel discovered definitly need addressing. I am not sure though, if this should be added to this PR (would require the consecutive pr's to be merged and I would add changes related to threads in a commit that is called "Add right panel chat timeline" which might be intuitive?
@jryans should we open new issues/pr's or should I add those features here?

**Review by gaelledel — CHANGES_REQUESTED** [review]
I am hovering on a thread within the chat panel and still cannot see the contextual menu with the "copy widget link" option
Screenshot 2021-11-29 at 13 53 03

**Comment by gaelledel** [comment]
See the desired result below
Screenshot 2021-11-29 at 13 57 50

**Comment by jryans** [comment]
> The issues @gaelledel discovered definitly need addressing. I am not sure though, if this should be added to this PR (would require the consecutive pr's to be merged and I would add changes related to threads in a commit that is called "Add right panel chat timeline" which might be intuitive? @jryans should we open new issues/pr's or should I add those features here?

Since this area does overlap with threads work, I agree with @toger5's instinct that we should capture this feedback in issues for the project and improve it in follow up PRs. They should be fast to review as well (both code and design wise) as they'll likely be smaller tweaks in a focused area.

**Comment by toger5** [comment]
The contextual menu now hides/shows dependent on the content of the main split area.
If there is the timeline the option to view the thread in the timeline appears


The same is done for the button on the room header



This is to clarify my changes. 
Those to changes were necessary anyways since the button makes no sense if the main timeline is not shown.
Adding the "copy link to thread" option to the context menu of the right panel timeline could be done as well.

**Review by gaelledel — APPROVED** [review]
